### PR TITLE
chore(deps): bump keyring controller to ^15.0.0

### LINKED ISSUE #21461: [Sentry] Error: Identity for 'undefined not found
Sentry Issue: [METAMASK-X7D1](https://metamask.sentry.io/issues/4539859147/?referrer=github_integration)

```
Error: Identity for 'undefined not found
  at exports.default.setSelectedAddress (app/scripts/controllers/preferences.js:405:13)
  at MetamaskController.addNewAccount (app/scripts/metamask-controller.js:3156:33)
  at async s.<anonymous> (app/scripts/lib/createMetaRPCHandler.js:23:7)
```

### LINKED ISSUE #23036: [Bug]: P0 - Import account through private key, i get two of identical account
## Version
- MacOS M1 Pro 13.5
- MetaMask 11.7.5

## What happened?
One day, I uninstall the Chrome Broswer,  for a while time later, i re-install the Chrome, so i should re-install MetaMask Extension, fortunately, i have private key on MetaMask Mobile App, I copy the private key and import to web MetaMask Extension, but i get the below case(see image

Other than that, I can't add a new account anymore(besides import method

image

### Expected behavior

1. Remove the repeat account
2. Can add new account whenever

### Screenshots/Recordings

_No response_

### Steps to reproduce

see the above desc

### Error messages or log output

_No response_

### Version

11.7.5

### Build type

None

### Browser

Chrome

### Operating system

MacOS

### Hardware wallet

_No response_

### Additional context

_No response_

### Severity

_No response_

### PR DESCRIPTION
**Pull request by montelaidev**
## **Description**

This PR bumps the `metamask/keyring-controller` to `^15.0.0` and addresses a bug in the keyring controller where invoking addNewAccount inadvertently retrieves addresses from all keyrings, rather than exclusively from the HD keyring. This behavior leads to inconsistencies and errors when accounts from different keyring types are present. The primary symptom of this bug is the retrieval of an incorrect number of accounts when adding a new HD account. When the total number of accounts in the client exceeds the number of HD accounts, it would cause the retrieval of an account that does not exist yet on the HD keyring.

## **Related issues**

Fixes: [23036](https://github.com/MetaMask/metamask-extension/issues/23036)
Fixes: [21461](https://github.com/MetaMask/metamask-extension/issues/21461)

## **Manual testing steps**

1. Import 2 private key accounts
2. Add a new HD account and see that it is now successfully added. 

## **Screenshots/Recordings**

Not applicable

## **Pre-merge author checklist**

- [x] I’ve followed [MetaMask Coding Standards](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/CODING_GUIDELINES.md).
- [x] I've completed the PR template to the best of my ability
- [ ] I’ve included tests if applicable
- [ ] I’ve documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [x] I’ve applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.

### CODE REVIEW

**Comment by montelaidev** [comment]
@metamaskbot update-policies

**Comment by montelaidev** [comment]
> Does this relate to [MetaMask/core#4203](https://github.com/MetaMask/core/pull/4203) in any way?

Yes it is related but the extension right now does not use `addNewAccountForKeyring` to create accounts, it only uses `addNewAccount` to create accounts.

File: ui/store/actions.ts [hunk 1]
```diff
@@ -454,7 +455,10 @@ export function addNewAccount(): ThunkAction<
 > {
   log.debug(`background.addNewAccount`);
   return async (dispatch, getState) => {
-    const oldAccounts = getInternalAccounts(getState());
+    const oldAccounts = getInternalAccounts(getState()).filter(
```
Comment (ccharly) [inline ui/store/actions.ts:458]: I wonder if we could have something like `getInternalAccountsForKeyring(getState(), KeyringTypes.hd)`?

File: package.json [hunk 2]
```diff
@@ -240,7 +240,9 @@
...
     "@metamask/network-controller@npm:^17.2.1": "patch:@metamask/network-controller@npm%3A18.1.0#~/.yarn/patches/@metamask-network-controller-npm-18.1.0-680908c29a.patch",
     "@metamask/network-controller@npm:^18.1.0": "patch:@metamask/network-controller@npm%3A18.1.0#~/.yarn/patches/@metamask-network-controller-npm-18.1.0-680908c29a.patch",
-    "@metamask/keyring-controller@npm:^13.0.0": "patch:@metamask/keyring-controller@npm%3A13.0.0#~/.yarn/patches/@metamask-keyring-controller-npm-13.0.0-d94816a680.patch",
+    "@metamask/keyring-controller@npm:^13.0.0": "patch:@metamask/keyring-controller@npm%3A15.0.0#~/.yarn/patches/@metamask-keyring-controller-npm-15.0.0-fa070ce311.patch",
+    "@metamask/keyring-controller@npm:^12.2.0": "patch:@metamask/keyring-controller@npm%3A15.0.0#~/.yarn/patches/@metamask-keyring-controller-npm-15.0.0-fa070ce311.patch",
+    "@metamask/keyring-controller@npm:^14.0.1": "patch:@metamask/keyring-controller@npm%3A15.0.0#~/.yarn/patches/@metamask-keyring-controller-npm-15.0.0-fa070ce311.patch",
```
Comment (gantunesr) [inline package.json:245]: Why is this adding patches for versions 12 and 13?

File: ui/store/actions.ts (hunk 1 above)
Comment (gantunesr) [reply ui/store/actions.ts:458]: Maybe you can even make it more generic to filter by different parameters, think that a lot of different accounts are going to be handled by the `eth-snap-keyring`

File: .yarn/patches/@metamask-keyring-controller-npm-15.0.0-fa070ce311.patch [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
++     * The HD keyring is the default keyring for all wallets if this keyring is missing
++     * for some reason we should avoid saving the keyrings
++     *
++     * The upstream fix is here: https://github.com/MetaMask/core/pull/4168
++     *
++     * This patch can be found on the core branch `extension-keyring-controller-v13-patch`
++     */
++    if (!serializedKeyrings.some((keyring) => keyring.type === KeyringTypes.hd)) {
++      throw new Error(KeyringControllerError.NoHdKeyring);
++    }
```
Comment (gantunesr) [inline .yarn/patches/@metamask-keyring-controller-npm-15.0.0-fa070ce311.patch:30]: I have two questions about the patch,

1. Are you adding anything new?
2. Have you checked if everything inside the patch is still necessary?

File: ui/store/actions.ts (hunk 1 above)
Comment (gantunesr) [reply ui/store/actions.ts:458]: This is, of course, out of scope for this PR

File: yarn.lock [hunk 4]
```diff
@@ -5460,46 +5444,7 @@ __metadata:
...
-    "@metamask/message-manager": "npm:^7.3.9"
-    "@metamask/utils": "npm:^8.3.0"
-    async-mutex: "npm:^0.2.6"
-    ethereumjs-wallet: "npm:^1.0.1"
-    immer: "npm:^9.0.6"
-  checksum: 477b6d7b9104370f562a862a11406013a744109f8d49e34eacedef8c9a1101e9f02f101aca3ea3af2b32b77b17535443c02d771c384ef436718b5d2fc28cbbb7
-  languageName: node
-  linkType: hard
-
-"@metamask/keyring-controller@npm:^12.2.0":
```
Comment (legobeat) [inline yarn.lock:5484]: Note that the package dependency `@metamask/signature-controller@^12.0.0` has a peerDependency on `@metamask/keyring-controller@^12.2.0`, which this breaks. Has it been verified that the two are compatible, or does `@metamask/signature-controller` need to be upgraded together with `@metamask/keyring-controller`?

Possibly related through also being peerDependencies of `@metamask/signature-controller`:
- #24141
- #24143

File: package.json (hunk 2 above)
Comment (legobeat) [reply package.json:245]: Note that the package dependency `@metamask/signature-controller@^12.0.0` has a peerDependency on `@metamask/keyring-controller@^12.2.0`, which this breaks. Has it been verified that the two are compatible, or does `@metamask/signature-controller` need to be upgraded together with `@metamask/keyring-controller`?

Possibly related through also being peerDependencies of `@metamask/signature-controller`:
- #24141
- #24143

Comment (gantunesr) [reply package.json:245]: It looks like both work together @legobeat

Comment (montelaidev) [reply package.json:245]: These resolutions were automatically made when patching. I've tested the signing with an eoa on the test dapp and the e2e are also passing.

File: ui/store/actions.ts (hunk 1 above)
Comment (montelaidev) [reply ui/store/actions.ts:458]: Yes, that makes sense. I'll create this in another pr

Comment (gantunesr) [reply ui/store/actions.ts:458]: Sounds good!

**Review by gantunesr — APPROVED** [review]
lgtm

**Review by ccharly — APPROVED** [review]

**Pull request merged**